adapters/logan/logan.go

```diff
@@ -0,0 +1,147 @@
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+type reqBodyExt struct {
+	LoganBidderExt reqBodyExtBidder `json:"bidder"`
+}
+
+type reqBodyExtBidder struct {
+	Type        string `json:"type"`
+	PlacementID string `json:"placementId,omitempty"`
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var err error
+	var adapterRequests []*adapters.RequestData
+
+	reqCopy := *request
+	for _, imp := range request.Imp {
+		reqCopy.Imp = []openrtb2.Imp{imp}
+
+		var bidderExt adapters.ExtImpBidder
+		var loganExt openrtb_ext.ImpExtLogan
+
+		if err = json.Unmarshal(reqCopy.Imp[0].Ext, &bidderExt); err != nil {
+			return nil, []error{err}
+		}
+		if err = json.Unmarshal(bidderExt.Bidder, &loganExt); err != nil {
+			return nil, []error{err}
+		}
+
+		temp := reqBodyExt{LoganBidderExt: reqBodyExtBidder{}}
+		temp.LoganBidderExt.PlacementID = loganExt.PlacementID
+		temp.LoganBidderExt.Type = "publisher"
+
+		finalyImpExt, err := json.Marshal(temp)
+		if err != nil {
+			return nil, []error{err}
+		}
+
+		reqCopy.Imp[0].Ext = finalyImpExt
+
+		adapterReq, err := a.makeRequest(&reqCopy)
+		if err != nil {
+			return nil, []error{err}
+		}
+
+		if adapterReq != nil {
+			adapterRequests = append(adapterRequests, adapterReq)
+		}
+	}
+	return adapterRequests, nil
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, error) {
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+	}, err
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for i := range seatBid.Bid {
+			bidType, err := getMediaTypeForImp(seatBid.Bid[i].ImpID, request.Imp)
```

Comment:
If the length of `seatBid.Bid` is `m` and the length of `request.Imp` is `n`, we have a time complexity of `O(mn)` with the current approach. At the cost of `O(n)`space we can improve the time complexity to `O(m+n)` if we map the `request.Imp[i].ID`s to the index values:
```
105       var response openrtb2.BidResponse
106       if err := json.Unmarshal(responseData.Body, &response); err != nil {
107           return nil, []error{err}
108       }
109  
110       bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
111       bidResponse.Currency = response.Cur
    +
    +     impsMappedByID := make(map[string]int, len(request.Imp))
    +     for i := range request.Imp {
    +         impsMappedByID[request.Imp[i].ID] = i
    +     }
    +
112       for _, seatBid := range response.SeatBid {
113           for i := range seatBid.Bid {
114 -             bidType, err := getMediaTypeForImp(seatBid.Bid[i].ImpID, request.Imp)
    +             bidType, err := getMediaTypeForImp(seatBid.Bid[i].ImpID, request.Imp, impsMappedByID)
115               if err != nil {
116                   return nil, []error{err}
117               }
118  
119               b := &adapters.TypedBid{
120                   Bid:     &seatBid.Bid[i],
121                   BidType: bidType,
122               }
123               bidResponse.Bids = append(bidResponse.Bids, b)
124           }
125       }
126       return bidResponse, nil
127   }
128  
129 - func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
    + func getMediaTypeForImp(impID string, imps []openrtb2.Imp, impMap map[string]int) (openrtb_ext.BidType, error) {
130 -     for _, imp := range imps {
131 -         if imp.ID == impID {
    +     if index, found := impMap[impID]; found {
132               if imp.Banner != nil {
133                   return openrtb_ext.BidTypeBanner, nil
134               }
135               if imp.Video != nil {
136                   return openrtb_ext.BidTypeVideo, nil
137               }
138               if imp.Native != nil {
139                   return openrtb_ext.BidTypeNative, nil
140               }
141 -         }
142       }
143  
144       return "", &errortypes.BadInput{
145           Message: fmt.Sprintf("Failed to find impression \"%s\"", impID),
146       }
147   }
adapters/logan/logan.go
```